test/src/test/java/org/corfudb/runtime/collections/CorfuStoreShimTest.java

```diff
@@ -683,4 +683,45 @@ public void queueOrderInTxnContext() throws Exception {
         }
         log.debug(corfuQueue.getMetrics().toString());
     }
+    /**
+     * ProtobufDescriptorTable should de-duplicate common protobuf file descriptors.
+     * This test creates a large number of tables that share protobuf files and validates
+     * that de-duplication occurs.
+     * Also exercises schema change validation logic.
+     * @throws Exception exception
+     */
+    @Test
+    public void checkProtoDeduplication() throws Exception {
+        CorfuRuntime corfuRuntime = getTestRuntime();
+        CorfuStoreShim shimStore = new CorfuStoreShim(corfuRuntime);
+        final String someNamespace = "some-namespace";
+        final String tableNamePrefix = "prefixTable";
+        final int numTables = PARAMETERS.NUM_ITERATIONS_VERY_LOW;
+
+        for (int i = 0; i < numTables; i++) {
+            shimStore.openTable(
+                    someNamespace,
+                    tableNamePrefix+i,
+                    UuidMsg.class,
+                    ManagedMetadata.class,
+                    ManagedMetadata.class,
+                    // TableOptions includes option to choose - Memory/Disk based corfu table.
+                    TableOptions.builder().build());
+        }
+        // Now update the schemas with different protobuf files
+        final int numProtoFiles = corfuRuntime.getTableRegistry().getProtobufDescriptorTable().size();
+        assertThat(numProtoFiles).isLessThan(numTables);
+        for (int i = 0; i < numTables; i++) {
+            shimStore.openTable(
+                    someNamespace,
+                    tableNamePrefix+i,
+                    UuidMsg.class,
+                    org.corfudb.runtime.proto.LogData.LogDataMsg.class, // this brings in 1 new protobuf file
+                    ManagedMetadata.class,
+                    // TableOptions includes option to choose - Memory/Disk based corfu table.
+                    TableOptions.builder().build());
+        }
+        final int numProtoFiles2 = corfuRuntime.getTableRegistry().getProtobufDescriptorTable().size();
+        assertThat(numProtoFiles2).isEqualTo(numProtoFiles + 1);
```

Comment:
This assertion doesn't really check for duplicates. Can you build a `CorfuTable<ProtobufFileName, CorfuRecord<ProtobufFileDescriptor, TableMetadata>>` and make sure its equal to `corfuRuntime.getTableRegistry().getProtobufDescriptorTable()`

Reply:
The test updates the schema in 10 tables but the number of descriptors updated is only 1. How is that not de-duplication?
What is a new CorfuTable going to demonstrate that the above test does not?

Reply:
Meant to say another map with the expected mapping and not a `CorfuTable`. 

> assertThat(numProtoFiles2).isEqualTo(numProtoFiles + 1);

Is not a sufficient test, what if two keys are being mapped to the same value ?

Reply:
De-duplicating actual FileDescriptorProto contents is not trivial and beyond the scope of this change.
This PR only seeks to store one FileDescriptorProto per filename.

Reply:
Instead of constantly trying to explain what the code "should" do in the comments section its a better idea to enforce it in code with a **properly** written **unit test**. In this particular test case merely checking the size of the map doesn't assert your expectations of the map's contents. Iterating the mapping one-by-one and assert that the contents map the right file name (ProtobufFileName) to the correct descriptor.

Reply:
Sure, will do. But please note that it's really not possible for the map to have incorrect contents and still have all the CorfuStoreIT, StreamingIT and CorfuStoreBrowserIT pass, because the new map is critical to build the dynamic message.